logstash-core/src/main/java/org/logstash/ackedqueue/Page.java

```diff
@@ -32,30 +33,52 @@ public Page(int pageNum, Queue queue, long minSeqNum, int elementCount, long fir
         this.ackedSeqNums = ackedSeqNums;
         this.lastCheckpoint = new Checkpoint(0, 0, 0, 0, 0);
         this.pageIO = pageIO;
+        this.writable = writable;
     }
 
     public String toString() {
         return "pageNum=" + this.pageNum + ", minSeqNum=" + this.minSeqNum + ", elementCount=" + this.elementCount + ", firstUnreadSeqNum=" + this.firstUnreadSeqNum;
     }
 
     /**
-     * @param limit the maximum number of elements to read
-     * @return {@link SequencedList} collection of elements read. the number of elements can be {@literal <=} limit
+     * @param limit the maximum number of elements to read, actual number readcan be smaller
+     * @return {@link SequencedList} collection of serialized elements read
+     * @throws IOException
      */
     public SequencedList<byte[]> read(int limit) throws IOException {
-
         // first make sure this page is activated, activating previously activated is harmless
         this.pageIO.activate();
 
         SequencedList<byte[]> serialized = this.pageIO.read(this.firstUnreadSeqNum, limit);
         assert serialized.getSeqNums().get(0) == this.firstUnreadSeqNum :
-            String.format("firstUnreadSeqNum=%d != first result seqNum=%d", this.firstUnreadSeqNum, serialized.getSeqNums().get(0));
+                String.format("firstUnreadSeqNum=%d != first result seqNum=%d", this.firstUnreadSeqNum, serialized.getSeqNums().get(0));
 
         this.firstUnreadSeqNum += serialized.getElements().size();
 
         return serialized;
     }
 
+    public void write(byte[] bytes, long seqNum, int checkpointMaxWrites) throws IOException {
+        if (! this.writable) {
+            throw new IOException(String.format("page=%d is not writable", this.pageNum));
```

Comment:
@colinsurprenant This isn't an `IOException` is it? We're throwing this as a result of a logic error (trying to write to a tail page) as far as I can tell. If so we must throw a `RuntimeException` like `IllegalStateException`.